packages/angular/ssr/node/src/app-engine.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { ɵAngularAppEngine as AngularAppEngine } from '@angular/ssr';
+import type { IncomingMessage } from 'http';
+import { createWebRequestFromNodeRequest } from './request';
+
+/**
+ * Angular server application engine.
+ * Manages Angular server applications (including localized ones) and handles rendering requests.
+
+ * @developerPreview
+ */
+export interface AngularNodeServerAppManager {
+  /**
+   * Renders an HTTP response based on the incoming request using the Angular server application.
+   *
+   * The method processes the incoming request, determines the appropriate route, and prepares the
+   * rendering context to generate a response. If the request URL corresponds to a static file (excluding `/index.html`),
+   * the method returns `null`.
+   *
+   * Example: A request to `https://www.example.com/page/index.html` will render the Angular route
+   * associated with `https://www.example.com/page`.
+   *
+   * @param request - The incoming HTTP request object to be rendered. It can be a `Request` or `IncomingMessage` object.
+   * @param requestContext - Optional additional context for the request, such as metadata or custom settings.
+   * @returns A promise that resolves to a `Response` object, or `null` if the request URL is for a static file
+   * (e.g., `./logo.png`) rather than an application route.
+   */
+  render(request: Request | IncomingMessage, requestContext?: unknown): Promise<Response | null>;
+}
+
+/**
+ * Angular server application engine.
+ * Manages Angular server applications (including localized ones), handles rendering requests,
+ * and optionally transforms index HTML before rendering.
+ */
+export class AngularNodeAppEngine extends AngularAppEngine implements AngularNodeServerAppManager {
+  /**
+   * Renders an HTTP response based on the incoming request using the Angular server application.
+   *
+   * The method processes the incoming request, determines the appropriate route, and prepares the
+   * rendering context to generate a response. If the request URL corresponds to a static file (excluding `/index.html`),
+   * the method returns `null`.
+   *
+   * Example: A request to `https://www.example.com/page/index.html` will render the Angular route
+   * associated with `https://www.example.com/page`.
+   *
+   * @param request - The incoming HTTP request object to be rendered. It can be a `Request` or `IncomingMessage` object.
+   * @param requestContext - Optional additional context for the request, such as metadata or custom settings.
+   * @returns A promise that resolves to a `Response` object, or `null` if the request URL is for a static file
+   * (e.g., `./logo.png`) rather than an application route.
+   */
+  override render(
+    request: Request | IncomingMessage,
```

Comment:
Question: Is there a particular reason we want to support `Request` as an input? If you're handling `Request` objects, you can just use `AngularAppEngine` directly right?

Is this so we can leverage `extends AngularAppEngine` here by making the `request` parameter compatible? Should `AngularNodeAppEngine` extend `AngularAppEngine` like this if they have different inputs?

Reply:
Correct, `Request` was added as a type to support extension.

A potentially alternative approach might look like this:

```typescript
class AngularNodeAppEngine {
  private angularAppEngine = new AngularAppEngine();

  async render(request: IncomingMessage): Promise<Response> {
    return this.angularAppEngine.toWebRequest(request);
  }
}
```

WDYT?

Reply:
A composition approach LGTM. If these classes don't need to extend each other, I suspect it would be more flexible and maintainable to keep their class hierarchies distinct.